<a href="https://colab.research.google.com/github/ViniciusARZ/GoogleSpreadsheets_DailyUpdate_ETL.ipynb/blob/main/BR_S%C3%A3o_Carlos'_Junior_Enterprises_Nucleos_2021_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Author: Vinícius Antoino Ramos Zecca

*LinkedIn: /viniciusarz*

*Github: /ViniciusARZ*


In [6]:
import sqlite3
import pandas as pd
from IPython.display import display
import sqlalchemy


# Data Cleaning and Processing (w/ Tools)

## Google Spreadsheets

In Google Spreadsheets I cleaned up what it was faster than it would be using SQL:
1. Replace "N/A's with " ", or basically *nulls*
2. Transform some Attribute types to make it easier to read in SQL.
3. Column "Faturamento" was in Text but it actually is a currency, so I used SPLIT() + TRIM() + SUBSTITUTE() + 'Number Formatting' to correct it... Same thing  with others.
4. Select which Columns I'm going to need based on the problem I want to solve and the analysis object. Did a backup in case I wanted to return. 


## SQL - Google Cloud's Big Query

Disclaimers:
1. I didn't use SQLite extension because I already had everything in Big Query
2. The following Query Codes are just to show what I did in SQL and keep a Changelog




Import and Schema Establishment

In [24]:
#'ID:INTEGER,EMPRESA_JUNIOR:STRING,DATA_DE_CADASTRO:DATE,ANO_DE_FEDERACAO:INTEGER, CLUSTER_2020:INTEGER,CLUSTER_2021:INTEGER,CLUSTER_2022:INTEGER,É_ALTO_CRESCIMENTO:BOOLEAN, MES_DO_ALTO_CRESCIMENTO:INTEGER,TEMPO_MEDIO_POR_PROJETO_DIAS:FLOAT,PORC_PROJETOS_COLETADOS_NPS:INTEGER,DATA_DE_ASSINATURA:DATE,MES:INTEGER,PROJETOS:INTEGER,DURACAO_DO_PROJETO_EM_DIAS_UTEIS:INTEGER,FATURAMENTO:FLOAT,ATINGE_ALGUM_ODS:BOOLEAN,NUMERO_DE_ODS_ATINGIDOS:INTEGER,ACAO_COMPARTILHADA:BOOLEAN,NUMERO_DE_EJS_QUE_PARTICIPARAM:INTEGER,NUMERO_DE_MEMBROS_PARTICIPANTES:INTEGER,NPS:INTEGER,DATA_DE_COLETA_DE_NPS:DATE,ESTADO_DO_CLIENTE:STRING,CIDADE_DO_CLIENTE:STRING,SERVICO:STRING,ESTADO:STRING,CIDADE:STRING,UNIVERSIDADE:STRING,CURSOS:STRING,NUMERO_MEMBROS:INTEGER'

# Dataframe reading just to visualize the table
df2021 = pd.read_csv('/content/BJ_NuSC_Mon2021_07 - 2021.csv',',')
display(df2021)


ID  ... MEMBROS
0     100  ...      69
1     100  ...      69
2     100  ...      69
3     100  ...      69
4     100  ...      69
..    ...  ...     ...
440  2904  ...      14
441  2904  ...      14
442  2904  ...      14
443  2904  ...      14
444  2904  ...      14

[445 rows x 31 columns]

*The SQL Queries to create Variables are in the Analysis Section below*

# Analysis Phase

In [27]:
# Aggregate data by Junior Enterprise and tell me the Cluster progress rate,
#   so this means that I want to see the proportion between Junior Enterprises that
#   were able to progress from 2020 to 2021 in Cluster classification.
#   QUERY 1
#   SELECT EMPRESA_J__NIOR,
    #   AVG(CLUSTER_2020) AS CLUSTER2020,
    #   AVG(CLUSTER_2021)AS CLUSTER2021,
    #   FROM `dataset-orbe.IRHistoricalAnalysis.NUSC_2021_07`
    #   GROUP BY EMPRESA_J__NIOR
    #   ORDER BY C2021 DESC
resultsQuery1 = pd.read_csv('results-query1.csv')
display(resultsQuery1)

# What can be seen from there?

# Cluster Evolution
# - In SQL:
#   SELECT EMPRESA_J__NIOR,
    #   AVG(CLUSTER_2020) AS CLUSTER2020,
    #   AVG(CLUSTER_2021)AS CLUSTER2021,
    #     ( SELECT COUNT(ID), 
    #     FROM `dataset-orbe.IRHistoricalAnalysis.NUSC_2021_07` 
    #     WHERE CLUSTER_2021 != CLUSTER_2020
    #     ) AS Percent
    #   FROM `dataset-orbe.IRHistoricalAnalysis.NUSC_2021_07`
    #   GROUP BY EMPRESA_J__NIOR
    #   ORDER BY CLUSTER2021 DESC
# In pandas:
NoJEProgressedCluster = resultsQuery1[resultsQuery1['CLUSTER2020'] != resultsQuery1['CLUSTER2021']]
display(NoJEProgressedCluster.count()/resultsQuery1.count())

# Cluster Proportion
# QUERY 2
#   SELECT CLUSTER_2020, 
    #   count(CLUSTER_2021) as Cluster2021_count,
    #   count(CLUSTER_2021) * 100.0 / (
    #       select count(*) from `dataset-orbe.IRHistoricalAnalysis.NUSC_2021_07`) 
    #       AS Cluster_percent2021
    #   FROM `dataset-orbe.IRHistoricalAnalysis.NUSC_2021_07`
    #   group by CLUSTER_2020

ID  CLUSTER2020  CLUSTER2021
0    100          5.0          5.0
1    163          4.0          5.0
2    421          3.0          5.0
3    137          4.0          5.0
4    400          4.0          4.0
5    302          1.0          4.0
6    131          3.0          4.0
7    142          4.0          4.0
8    168          3.0          4.0
9   1444          1.0          3.0
10   565          4.0          3.0
11   143          2.0          3.0
12   303          3.0          3.0
13   723          3.0          3.0
14   146          1.0          3.0
15   147          2.0          2.0
16   405          3.0          2.0
17   434          2.0          2.0
18   543          1.0          2.0
19   840          2.0          2.0
20  1395          2.0          2.0
21   607          1.0          2.0
22   144          2.0          2.0
23   692          4.0          2.0
24  1407          1.0          2.0
25  1576          1.0          2.0
26  1121          1.0          1.0
27  1512          1.0          1.0
28  1692          1.0          1.0
29   834          1.0          1.0
30  1530          1.0          1.0
31   442          1.0          1.0
32  2495          NaN          1.0
33   724          1.0          1.0
34   167          1.0          1.0
35   882          1.0          1.0
36   406          1.0          1.0
37   691          1.0          1.0
38  1116          1.0          1.0
39  1506          1.0          1.0
40  1562          1.0          1.0
41  1582          1.0          1.0
42  1745          1.0          1.0
43   910          1.0          1.0
44  1342          1.0          1.0
45  1416          1.0          1.0
46  1511          1.0          1.0
47  1616          1.0          1.0
48  1735          1.0          1.0
49  2904          NaN          NaN

ID             0.360000
CLUSTER2020    0.333333
CLUSTER2021    0.346939
dtype: float64


If the ID is taken into account, because it is the only value that can't be *null*, this means that only 0.36, or **36%**, of Junior Enterprises were **able to Progress in Cluster Classification**